### ISSUE TITLE
Support for velocities (& forces) in trajectory I/O

### ISSUE THREAD

**Issue by salotz** [mdtraj/mdtraj#1824]
Currently in the low level APIs velocities and forces are not supported even if the file format supports them, e.g. trr and Amber restarts.

This should be supported at least at the low level. I would propose not changing the high-level APIs of `Trajectory` but just adding in support for it in each individual low level that supports it. To support backwards compatibility these extra fields could be only enabled via flags to the `read` and `write` methods.

**Comment by ctk3b** [comment]
We've had a couple of folks ask related questions. The responses include a few workarounds and some ideas for how we could implement this without a major change to the high-level APIs:

https://github.com/mdtraj/mdtraj/issues/1820
https://github.com/mdtraj/mdtraj/issues/1255
https://github.com/mdtraj/mdtraj/pull/1271
https://github.com/mdtraj/mdtraj/issues/1230
https://github.com/mdtraj/mdtraj/issues/36

But the TLDR is that there are no _active_ efforts or plans to implement support for velocities or forces

**Comment by salotz** [comment]
This is something I would consider make PRs for as demand requires.

This one could serve me as a stop-gap: https://github.com/mdtraj/mdtraj/pull/1294

Didn't notice that `_read` method. I still feel a better approach is to add the kwargs to the main `read` method. Its supposed to be "hidden" but its meant for use outside the library so its just messaging incorrectly. Would the opinion of the maintainers currently support just upgrading that `_read` and merging it with the `read` method? Or perhaps just create an alternative  public method like `read_all` if you don't want to touch `read`.

**Comment by sukritsingh** [comment]
> Would the opinion of the maintainers currently support just upgrading that _read and merging it with the read method? Or perhaps just create an alternative public method like read_all if you don't want to touch read.

I would lean towards the latter (an alternative `read_all` method rather than editing `read) - the reason is because there are some contexts like constructing MSMs or working with FAH datasets where it may not be useful to load all the velocities and forces into memory. The two options I see most compatible with current setup would be: 
1. Upgrading `_read` but ensuring it reads positions by defaults and only reads velocities and forces if explicitly specified.
2. Having a separate method for reading velocities and forces entirely.

**Comment by salotz** [comment]
@sukritsingh Just to clarify my proposal for reusing `read` would have them only returned when explicitly requested, e.g.

```python
def read(..., get_velocities=False, get_forces=False):
    ...
```

So none of the existing interface behavior would change. But I also think another method is acceptable.

I completely agree and support the low-level API giving you complete control over exactly what data to load. And in fact it could even be extended to all of the possible fields. For instance most of the low level readers differ on what fields they actually return like time, segment time, cell lengths/angle vs vectors, etc.

So maybe something like a new method:

```python
def read_fields(self, get_positions=True, get_time=False, ...):
```

I already end up doing this a lot in my application code using mdtraj and its probably something general enough to be useful to a lot of other people. On top of that it paves the way for a more generic trajectory file reader that doesn't require a topology which can be very annoying to have to provide if you just want the coordinates and have it auto-detect the file format.

**Comment by sukritsingh** [comment]
> So none of the existing interface behavior would change. 

My worry would be that you'd have modify the output of the `read` method if someone specified outputs that are different from the default set of `read` outputs. That kind of customization in the output may have an impact on a lot of automated code out there other folks are using. Perhaps the "long term solution" here is to have these kinds of things as part of the `Trajectory` object explicitly and stored as `None` unless specified in `read`? That would be a much larger effort, though.

I like the idea of a new method like `read_fields` or something like that!

**Comment by mattwthompson** [comment]
I don't see a downside to adding `Trajectory.velocities` or whatever, defaulting to and remaining `None` in 99.99% of use cases

**Comment by sukritsingh** [comment]
>  I don't think I would use it very often but I think its a fine addition, but probably a separate issue and PR.
Agreed – If this is something lots of folks want, definitely open to altering `read` to make it backwards compatible and adding `Trajectory.velocities`

> That said I would still use read_fields if given the option so I'm sensing that is the more acceptable choice.

That was how I was leaning as well for now but I don't have too strong feelings either way - feel free to submit a PR addressing this either way!

**Comment by helmutcarter** [comment]
This would be a very useful addition for my work, @salotz
